File: grammars/fixtures/sample.ex
defmodule Poly.Release do
  @moduledoc """
  Parse and rank release tags.
  """

  @semver ~r/^v?(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+)(?:-(?<pre>.+))?$/

  defstruct tag: nil, assets: [], prerelease?: false

  @type t :: %__MODULE__{tag: String.t(), assets: [String.t()], prerelease?: boolean()}

  @doc "Return `{major, minor, patch}` for `tag`, or `:error`."
  @spec version(String.t()) :: {:ok, {integer, integer, integer}} | :error
  def version(tag) do
    case Regex.named_captures(@semver, tag) do
      nil -> :error
      %{"major" => a, "minor" => b, "patch" => c} ->
        {:ok, {String.to_integer(a), String.to_integer(b), String.to_integer(c)}}
    end
  end

  def latest(releases) do
    releases
    |> Enum.reject(& &1.prerelease?)
    |> Enum.sort_by(fn %__MODULE__{tag: tag} ->
      case version(tag) do
        {:ok, v} -> v
        :error -> {0, 0, 0}
      end
    end)
    |> List.last()
  end

  defimpl String.Chars do
    def to_string(%Poly.Release{tag: tag, assets: assets}) do
      "#{tag} (#{length(assets)} assets)"
    end
  end
end

File: grammars/fixtures/sample.erl
-module(poly_release).

-export([version/1, latest/1, describe/1]).

-record(release, {tag :: binary(), assets = [] :: [binary()], prerelease = false :: boolean()}).

-type release() :: #release{}.

-define(SEMVER, "^v?([0-9]+)\\.([0-9]+)\\.([0-9]+)(?:-(.+))?$").

%% @doc Parse a semver tag into a {Major, Minor, Patch} tuple.
-spec version(binary()) -> {ok, {integer(), integer(), integer()}} | error.
version(Tag) ->
    case re:run(Tag, ?SEMVER, [{capture, all_but_first, list}]) of
        {match, [Major, Minor, Patch | _]} ->
            {ok, {list_to_integer(Major), list_to_integer(Minor), list_to_integer(Patch)}};
        nomatch ->
            error
    end.

-spec latest([release()]) -> release() | undefined.
latest(Releases) ->
    Stable = [R || R = #release{prerelease = false} <- Releases],
    case lists:sort(fun(A, B) -> version(A#release.tag) =< version(B#release.tag) end, Stable) of
        [] -> undefined;
        Sorted -> lists:last(Sorted)
    end.

describe(#release{tag = Tag, assets = Assets}) ->
    io_lib:format("~s (~p assets)", [Tag, length(Assets)]).
